server/scsynth/SC_PortAudio.cpp

```diff
@@ -311,17 +312,82 @@ bool SC_PortAudioDriver::DriverSetup(int* outNumSamples, double* outSampleRate)
     mDeviceInOut[0] = paNoDevice;
     mDeviceInOut[1] = paNoDevice;
     if (mWorld->hw->mInDeviceName)
-        GetPaDeviceFromName(mWorld->hw->mInDeviceName, mDeviceInOut);
+        GetPaDeviceFromName(mWorld->hw->mInDeviceName, &mDeviceInOut[0], IOType::Input);
+    if (mWorld->hw->mOutDeviceName)
+        GetPaDeviceFromName(mWorld->hw->mOutDeviceName, &mDeviceInOut[1], IOType::Output);
+
+#ifdef _WIN32
+    // if one device is specified, let's try to open another one on matching api
+    if (mDeviceInOut[0] == paNoDevice && mDeviceInOut[1] != paNoDevice)
+        mDeviceInOut[0] = Pa_GetHostApiInfo(Pa_GetDeviceInfo(mDeviceInOut[1])->hostApi)->defaultInputDevice;
+    if (mDeviceInOut[0] != paNoDevice && mDeviceInOut[1] == paNoDevice)
+        mDeviceInOut[1] = Pa_GetHostApiInfo(Pa_GetDeviceInfo(mDeviceInOut[0])->hostApi)->defaultOutputDevice;
+    // check if devices are having mismatched API, but only if they are defined
+    if (mDeviceInOut[0] != paNoDevice && mDeviceInOut[1] != paNoDevice) {
+        if (Pa_GetDeviceInfo(mDeviceInOut[0])->hostApi != Pa_GetDeviceInfo(mDeviceInOut[1])->hostApi) {
+            fprintf(stdout, "Requested devices use different API. Setting output device to use %s\n",
+                    Pa_GetHostApiInfo(Pa_GetDeviceInfo(mDeviceInOut[0])->hostApi)->name);
+            mDeviceInOut[1] = Pa_GetHostApiInfo(Pa_GetDeviceInfo(mDeviceInOut[0])->hostApi)->defaultOutputDevice;
+        }
+    }
+    // in case we don't have a proper device, let's try to open one on ASIO
+    if (mDeviceInOut[0] == paNoDevice || mDeviceInOut[1] == paNoDevice) {
+        mDeviceInOut[0] = Pa_GetHostApiInfo(Pa_HostApiTypeIdToHostApiIndex(paASIO))->defaultInputDevice;
+        mDeviceInOut[1] = Pa_GetHostApiInfo(Pa_HostApiTypeIdToHostApiIndex(paASIO))->defaultOutputDevice;
+    }
+    // check for sample rate on  WASAPI devices
+    if (Pa_GetDeviceInfo(mDeviceInOut[0])->hostApi == Pa_HostApiTypeIdToHostApiIndex(paWASAPI)
```

Comment:
Here we get a crash if mDeviceInOut is still paNoDevice (happens is ASIO not present and a bad device is given)